core/metrics-core-service/src/main/antlr4/org/hawkular/metrics/core/service/tags/parser/TagQuery.g4

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2014-2017 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+grammar TagQuery;
+
+@header {
+/*
+ * Copyright 2014-2017 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+}
+
+
+//lexer
+tagquery
+  : object
+  ;
+
+object
+  : pair
+  | '(' object ')'
+  | object logical_operator object
+  ;
+
+pair
+  : key boolean_operator value
+  | key array_operator array
+  ;
+
+logical_operator
+  : AND
+  | OR
+  ;
+
+boolean_operator
+  : EQUAL
+  | NOTEQUAL
+  ;
+
+array_operator
+  : IN
+  | NOTIN
```

Comment:
I suggest to maintain the NOTIN separated to simplify the management of NOT operator.
I referenced this here, too.
https://github.com/hawkular/hawkular-alerts/pull/282#issuecomment-274758744